portal/src/graphql/portal/DNSConfigurationScreen.tsx

```diff
@@ -0,0 +1,598 @@
+        // otherwise assume https scheme
+        return `https://${domain.domain}`;
+      }
+    });
+    if (initialPublicOrigin !== "" && !keys.includes(initialPublicOrigin)) {
+      keys.unshift(initialPublicOrigin);
+    }
+    return keys;
+  }, [verifiedDomains, initialPublicOrigin]);
+
+  const {
+    options: publicOriginOptions,
+    selectedKey: publicOrigin,
+    onChange: onPublicOriginChange,
+  } = useDropdown(publicOriginOptionKeys, initialPublicOrigin);
+
+  const isModified = useMemo(() => {
+    return initialPublicOrigin !== publicOrigin;
+  }, [initialPublicOrigin, publicOrigin]);
+
+  const onSaveClick = useCallback(() => {
+    if (rawAppConfig == null) {
+      return;
+    }
+
+    const newAppConfig = produce(rawAppConfig, (draftConfig) => {
+      const newPublicOrigin =
+        publicOrigin?.trim() !== "" ? publicOrigin : undefined;
+      draftConfig.http = draftConfig.http ?? {};
+      draftConfig.http.public_origin = newPublicOrigin;
+
+      clearEmptyObject(draftConfig);
+    });
+
+    updateAppConfig(newAppConfig).catch(() => {});
+  }, [publicOrigin, rawAppConfig, updateAppConfig]);
+
+  return (
+    <section className={styles.publicOrigin}>
+      {updateAppConfigError && <ShowError error={updateAppConfigError} />}
+      <Text
+        as="h2"
+        className={cn(
+          styles.header,
+          styles.subHeader,
+          styles.publicOriginHeader
+        )}
+      >
+        <FormattedMessage id="DNSConfigurationScreen.public-origin.header" />
+      </Text>
+      <div className={styles.publicOriginInput}>
+        <Dropdown
+          className={styles.publicOriginField}
+          options={publicOriginOptions}
+          selectedKey={publicOrigin}
+          onChange={onPublicOriginChange}
+        />
+        <ButtonWithLoading
+          className={styles.savePublicOriginButton}
+          disabled={!isModified}
+          labelId="save"
+          loadingLabelId="saving"
+          loading={updatingAppConfig}
+          onClick={onSaveClick}
+        />
+      </div>
+    </section>
+  );
+};
+
+const AddDomainSection: React.FC = function AddDomainSection() {
+  const { renderToString } = useContext(Context);
+  const { appID } = useParams();
+  const { value: newDomain, onChange: onNewDomainChange } = useTextField("");
+  const {
+    createDomain,
+    loading: creatingDomain,
+    error: createDomainError,
+  } = useCreateDomainMutation(appID);
+
+  const onAddClick = useCallback(
+    (ev: React.FormEvent) => {
+      ev.preventDefault();
+      ev.stopPropagation();
+
+      createDomain(newDomain)
+        .then((success) => {
+          if (success) {
+            onNewDomainChange(null, "");
+          }
+        })
+        .catch(() => {});
+    },
+    [createDomain, newDomain, onNewDomainChange]
+  );
+
+  const isModified = useMemo(() => {
+    return newDomain !== "";
+  }, [newDomain]);
+
+  // NOTE: if domain is invalid, generic error without extensions is thrown
```

Comment:
This should be a bug in backend, I'll fix it to produce a proper API error when invalid domain is entered.